File: 3152-special-array-ii/3152-special-array-ii.js
var isArraySpecial = function(nums, queries) {
    const n = nums.length;

    for (let i = 1; i < n; i++) {
        if ((nums[i] & 1) === (nums[i - 1] & 1)) {
            nums[i - 1] = 1;
        } else {
            nums[i - 1] = 0;
        }
    }

    // Compute prefix sums
    for (let i = 1; i < n; i++) {
        nums[i] += nums[i - 1];
    }

    // Process queries
    return queries.map(([l, r]) => {
        if (l === r) {
            return true;
        }
        const res = (r > 0 ? nums[r - 1] : nums[r]) - (l > 0 ? nums[l - 1] : 0);
        return res === 0;
    });
};

File: 3152-special-array-ii/3152-special-array-ii.py
class Solution:
    def isArraySpecial(self, nums: list[int], queries: list[list[int]]) -> list[bool]:
        n = len(nums)

        # Preprocess the nums array
        for i in range(1, n):
            if (nums[i] & 1) == (nums[i - 1] & 1):
                nums[i - 1] = 1
            else:
                nums[i - 1] = 0

        # Compute prefix sums
        for i in range(1, n):
            nums[i] += nums[i - 1]

        # Process queries
        ans = []
        for l, r in queries:
            if l == r:
                ans.append(True)
            else:
                res = (nums[r - 1] if r > 0 else nums[r]) - (nums[l - 1] if l > 0 else 0)
                ans.append(res == 0)

        return ans
